### PR TITLE
Add disposeHandlers

### PR DESCRIPTION
**Pull request by aeisenberg**
These functions assist with object disposal. They add custom behaviour
during disposal. The primary usage of disposalHandlers is during testing
where some objects should not be disposed in order to avoid testing
errors.

Additionally, move DisposableObject to the pure folder and create unit
tests for it.

This is an internal change and does not require a changelog entry.

When dispose handlers are used, the error log during tests is significantly quieter.

## Checklist

- [n/a] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [n/a] Issues have been created for any UI or other user-facing changes made by this pull request.
- [n/a] `@github/docs-content-dsp` has been cc'd in all issues for UI or other user-facing changes made by this pull request.

### CODE REVIEW

File: extensions/ql-vscode/src/pure/disposable-object.ts [hunk 1]
```diff
@@ -1,4 +1,11 @@
-import { Disposable } from 'vscode';
+
+// Avoid explicitly referencing Disposable type in vscode.
+// This file cannot have dependencies on the vscode API.
+interface Disposable {
+  dispose(): any;
```
Comment (alexet) [inline extensions/ql-vscode/src/pure/disposable-object.ts:5]: Is `unknown` better than `any` here.

File: extensions/ql-vscode/src/vscode-tests/run-integration-tests.ts [hunk 2]
```diff
@@ -112,18 +112,21 @@ function getLaunchArgs(dir: TestDir) {
   switch (dir) {
     case TestDir.NoWorksspace:
       return [
-        '--disable-extensions'
+        '--disable-extensions',
```
Comment (alexet) [inline extensions/ql-vscode/src/vscode-tests/run-integration-tests.ts:115]: These `--disable-gpu` lines aren't mentioned in the commit message or PR are these meant to be part of this PR?

**Review by alexet — APPROVED** [review]

File: extensions/ql-vscode/src/pure/disposable-object.ts (hunk 1 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/pure/disposable-object.ts:5]: The vscode api uses `any`. I was trying to copy that.

File: extensions/ql-vscode/src/vscode-tests/run-integration-tests.ts (hunk 2 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/vscode-tests/run-integration-tests.ts:115]: Ah...Let me update the commit message. `--disable-gpu` also removes some error messages from the logs.

File: extensions/ql-vscode/src/pure/disposable-object.ts [hunk 3]
```diff
@@ -40,21 +47,39 @@ export abstract class DisposableObject implements Disposable {
...
+   *      all items of a disposable.
+   */
+  public dispose(disposeHandler?: DisposeHandler) {
     if (this.tracked !== undefined) {
       for (const trackedObject of this.tracked.values()) {
-        trackedObject.dispose();
+        if (disposeHandler) {
+          disposeHandler(trackedObject);
+        } else {
+          trackedObject.dispose();
```
Comment (adityasharad) [inline extensions/ql-vscode/src/pure/disposable-object.ts:71]: We don't have a situation where we want to _both_ call the dispose handler and `trackedObject.dispose()`, right?

File: extensions/ql-vscode/src/query-history.ts [hunk 4]
```diff
@@ -13,10 +13,10 @@ import {
 import { logger } from './logging';
 import { URLSearchParams } from 'url';
 import { QueryServerClient } from './queryserver-client';
-import { DisposableObject } from './vscode-utils/disposable-object';
+import { DisposableObject } from './pure/disposable-object';
 import { commandRunner } from './commandRunner';
 
-/**
+/**./pure/disposable-object
```
Comment (adityasharad) [inline extensions/ql-vscode/src/query-history.ts:19]: Find-replace leftover :)

File: extensions/ql-vscode/src/vscode-tests/test-dispose-handler.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { Disposable } from 'vscode';
+import { DisposableObject } from '../pure/disposable-object';
+
+export function testDisposeHandler(disposable: any & Disposable) {
+  if (disposable.onDidExpandElement && disposable.onDidCollapseElement && disposable.reveal) {
+    // This looks like a treeViewer. Don't dispose
+    return;
+  }
+  if (disposable instanceof DisposableObject) {
+    disposable.dispose(testDisposeHandler);
```
Comment (adityasharad) [inline extensions/ql-vscode/src/vscode-tests/test-dispose-handler.ts:10]: Doesn't this loop between the call to `disposable.dispose(testDisposeHandler)` here and to `disposeHandler(trackedObject)` in `DisposableObject.dispose`?

**Review by adityasharad — COMMENTED** [review]
Good find and I appreciate the greater stability. Also nice not depending on the VS Code API for this. A few questions.

File: extensions/ql-vscode/src/pure/disposable-object.ts (hunk 3 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/pure/disposable-object.ts:71]: The point is that the dispose handler determines whether or not to call the actual dispose. So, your statement is correct.

File: extensions/ql-vscode/src/vscode-tests/test-dispose-handler.ts (hunk 5 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/vscode-tests/test-dispose-handler.ts:10]: No. If a dispose handler is used, the `dispose` method of the `DisposableObject` never calls `dispose` directly on its children. Rather, it gets delegated to the handler.

File: extensions/ql-vscode/src/query-history.ts (hunk 4 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/query-history.ts:19]: opps...will fix.

File: extensions/ql-vscode/src/vscode-tests/test-dispose-handler.ts (hunk 5 above)
Comment (adityasharad) [reply extensions/ql-vscode/src/vscode-tests/test-dispose-handler.ts:10]: Oh yes a `DisposableObject` has disposable _children_. Ok cool. Your tests will expose it if we break this.

**Review by adityasharad — APPROVED** [review]

**Pull request merged**